sink/datadog.go

```diff
@@ -141,13 +160,70 @@ func NewDatadog(monitorId string, opts, tags map[string]string, httpClient *http
 
 		c := datadog.NewConfiguration()
 		c.HTTPClient = httpClient
+		c.Compress = d.compress
 		metricsApi := datadogV2.NewMetricsApi(datadog.NewAPIClient(c))
 		d.metricsApi = metricsApi
 	}
 
 	return d, nil
 }
 
+func (s *Datadog) sendApi(ddCtx context.Context, dp []datadogV2.MetricSeries) error {
+	// Send metrics to Datadog. The Datadog client handles everything; we just pass
+	// it data points.
+	payload := datadogV2.MetricPayload{
+		Series: dp,
+	}
+	_, r, err := s.metricsApi.SubmitMetrics(ddCtx, payload, *datadogV2.NewSubmitMetricsOptionalParameters())
+	// Check to see if the request was too large
+	if err != nil {
+		if r.StatusCode == http.StatusRequestEntityTooLarge {
+			oldMaxMetricsPerRequest := s.maxMetricsPerRequest
+
+			// Sample one of the metrics and try caculating a new maxMetricsPerRequest
+			if pErr := s.updateMaxMetricsPerRequest(dp[0], len(dp)); pErr != nil {
+				return fmt.Errorf("Unable to compute maxMetricsPerRequest to properly send metrics: %s: %s", pErr, err)
+			}
+
+			// Make a local copy to avoid this value changing due to other threads
+			localMaxMetricsPerRequest := s.maxMetricsPerRequest
+
+			if (localMaxMetricsPerRequest == len(dp) && oldMaxMetricsPerRequest == localMaxMetricsPerRequest) || oldMaxMetricsPerRequest == localMaxMetricsPerRequest {
+				// maxMetricsPerRequest already matches the number of metrics we are trying to send so we can't retry any more
+				return fmt.Errorf("Unable to send metrics due to size: %s", err)
+			}
+
+			// Using the updated maxMetricsPerRequest value, partition the metrics into new chunks and send again
+			var pErr error
+			rangeStart := 0
+			var i int
+			for i = localMaxMetricsPerRequest; i < len(dp); i += localMaxMetricsPerRequest {
+				if err = s.sendApi(ddCtx, dp[rangeStart:i]); err != nil {
+					pErr = fmt.Errorf("%s\n%s", pErr, err)
+				}
+				rangeStart = i
+			}
+
+			if i-rangeStart > 0 {
+				if i > len(dp) {
+					i = len(dp)
+				}
+
+				if err = s.sendApi(ddCtx, dp[rangeStart:i]); err != nil {
+					pErr = fmt.Errorf("%s\n%s", pErr, err)
+				}
+			}
+
+			return pErr
+		}
+
+		blip.Debug("error sending data points to Datadog: %s", err)
+		blip.Debug("error sending data points to Datadog: Http Response - %s", r)
+	}
+
+	return err
```

Comment:
`return nil` because the `if` above already checked that it's not nil, so an explicit `return nil` is idiomatic to show that func has reached a successful end/return.

Reply:
By contrast, this kind of `return err` is used when a func returns the last return val of its last call without checking that return value.